Library/Homebrew/cmd/list.rb

```diff
@@ -1,233 +1,242 @@
+                            "This is the default when output is not to a terminal."
+        switch "-l",
+               description: "List formulae and/or casks in long format. " \
+                            "Has no effect when a formula or cask name is passed as an argument."
+        switch "-r",
+               description: "Reverse the order of the formulae and/or casks sort to list the oldest entries first. " \
+                            "Has no effect when a formula or cask name is passed as an argument."
+        switch "-t",
+               description: "Sort formulae and/or casks by time modified, listing most recently modified first. " \
+                            "Has no effect when a formula or cask name is passed as an argument."
+
+        conflicts "--formula", "--cask"
+        conflicts "--pinned", "--cask"
+        conflicts "--multiple", "--cask"
+        conflicts "--pinned", "--multiple"
+        ["-1", "-l", "-r", "-t"].each do |flag|
+          conflicts "--versions", flag
+          conflicts "--pinned", flag
+        end
+        ["--versions", "--pinned", "-l", "-r", "-t"].each do |flag|
+          conflicts "--full-name", flag
+        end
 
-      named_args [:installed_formula, :installed_cask]
-    end
-  end
+        named_args [:installed_formula, :installed_cask]
+      end
 
-  def self.list
-    args = list_args.parse
+      sig { override.void }
+      def run
+        if args.full_name?
+          unless args.cask?
+            formula_names = args.no_named? ? Formula.installed : args.named.to_resolved_formulae
+            full_formula_names = formula_names.map(&:full_name).sort(&tap_and_name_comparison)
+            full_formula_names = Formatter.columns(full_formula_names) unless args[:"1?"]
+            puts full_formula_names if full_formula_names.present?
+          end
+          if args.cask? || (!args.formula? && args.no_named?)
+            cask_names = if args.no_named?
+              Cask::Caskroom.casks
+            else
+              args.named.to_formulae_and_casks(only: :cask, method: :resolve)
+            end
+            # The cast is because `Keg`` does not define `full_name`
+            full_cask_names = T.cast(cask_names, T::Array[T.any(Formula, Cask::Cask)])
+                               .map(&:full_name).sort(&tap_and_name_comparison)
+            full_cask_names = Formatter.columns(full_cask_names) unless args[:"1?"]
+            puts full_cask_names if full_cask_names.present?
+          end
+        elsif args[:pinned?]
+          filtered_list
+        elsif args[:versions?]
+          filtered_list unless args.cask?
+          list_casks if args.cask? || (!args.formula? && !args[:multiple?] && args.no_named?)
+        elsif args.no_named?
+          ENV["CLICOLOR"] = nil
+
+          ls_args = []
+          ls_args << "-1" if args[:"1?"]
+          ls_args << "-l" if args[:l?]
+          ls_args << "-r" if args[:r?]
+          ls_args << "-t" if args[:t?]
```

Comment:
Feels weaker from a typechecking point of view and feels weird seeing this but still seeing `args.formula?` and `args.verbose?`

Reply:
I'd overlooked this change. Is this needed to make Sorbet happy?

Reply:
Correct. Because `CLI::Args` subclasses `OpenStruct`, many of the commands are just hash accessors, and not supported by Sorbet. Thus `args`, as the result of `Parser#parse`, has been treated as [untyped](https://github.com/Homebrew/brew/blob/ea2892f8/Library/Homebrew/cli/parser.rb#L311-L314):
```ruby
      # @return [Args] The actual return type is `Args`, but since `Args` uses `method_missing` to handle options, the
      #   `sig` annotates this as returning `T.untyped` to avoid spurious type errors.
      sig { params(argv: T::Array[String], ignore_invalid_options: T::Boolean).returns(T.untyped) }
      def parse(argv = ARGV.freeze, ignore_invalid_options: false)
```

As part of this new implementation, args is correctly typed, but may require changes:
```ruby
    # @note because `Args` makes use `OpenStruct`, subclasses may need to use a tapioca compiler,
    #   hash accessors, args.rbi, or other means to make this work with legacy commands:
    sig { returns(CLI::Args) }
    attr_reader :args
```
(Reviewers may have noticed that `AbstractCommand` has the rare `# typed: strong`, which is often overkill, but helpful for ancestors or other extensible tooling.)

Some `args` commands still work, for various reasons. As noted above, `args.rbi` is one option, but i don't especially love it (and I don't really have the bandwidth to write type signatures as part of the first pass of converting commands). It makes the methods available to all commands, rather than for just the command(s) that define the flag. I'm actively thinking about how to namespace the args as well, as a sort of v2 of what I'm introducing here, but I don't have anything concrete to propose at this time.

Reply:
The problem here is that while `args.pinned?` and `args[:pinned?]` are equivalent the new approach doesn't handle typos well. `args.pinnedd?` raises an error `Error: CLI arg for `pinnedd?` is not declared for this command` while `args[:pinnedd?]` doesn't fail and just returns nil.

Could we just keep cli args untyped and avoid this change? What additional type safety do we get by typing this attribute?

Reply:
Also, this is just another example of why a refactor of `CLI::Args` might be in the cards at some point.

Reply:
> The problem here is that while `args.pinned?` and `args[:pinned?]` are equivalent the new approach doesn't handle typos well. `args.pinnedd?` raises an error `Error: CLI arg for `pinnedd?` is not declared for this command` while `args[:pinnedd?]` doesn't fail and just returns nil.
> 
> Could we just keep cli args untyped and avoid this change? What additional type safety do we get by typing this attribute?

Correct, though it still represents a marginal type improvement anywhere this change is made (in that sorbet now recognizes the type of `args`). This means that `args.pinnedd?` at this call site previously would have passed typecheck. Again, there are alternatives if folks find this unacceptable (tapioca and args.rbi, in particular), or we could punt to when we want to refactor `CLI::Args` more broadly (though I'm partial to avoiding `T.untyped` in `AbstractCommand` from the outset).

Reply:
Sounds like we've moving from something that checked for errors at runtime so something that never checks for errors?

Reply:
> Sounds like we've moving from something that checked for errors at runtime so something that never checks for errors?

Could you elaborate, possibly with examples? I'm not sure that I understand this claim.

>  I'm not entirely sure what the marginal type improvement here is.

`args` was previously untyped by default, but now is typed. This is an issue accessing when OpenStruct hash values via methods.

Reply:
```diff
diff --git a/Library/Homebrew/cmd/list.rb b/Library/Homebrew/cmd/list.rb
index 238f63f53a..d3f7953824 100644
--- a/Library/Homebrew/cmd/list.rb
+++ b/Library/Homebrew/cmd/list.rb
@@ -96,7 +96,7 @@ module Homebrew
 
           ls_args = []
           ls_args << "-1" if args[:"1?"]
-          ls_args << "-l" if args[:l?]
+          ls_args << "-l" if args.somethinginvalid?
           ls_args << "-r" if args[:r?]
           ls_args << "-t" if args[:t?]
```

results in:

```console
$ brew list
Error: CLI arg for `somethinginvalid?` is not declared for this command
```

---

However this:

```diff
diff --git a/Library/Homebrew/cmd/list.rb b/Library/Homebrew/cmd/list.rb
index 238f63f53a..f1fee19c14 100644
--- a/Library/Homebrew/cmd/list.rb
+++ b/Library/Homebrew/cmd/list.rb
@@ -96,7 +96,7 @@ module Homebrew
 
           ls_args = []
           ls_args << "-1" if args[:"1?"]
-          ls_args << "-l" if args[:l?]
+          ls_args << "-l" if args[:somethinginvalid?]
           ls_args << "-r" if args[:r?]
           ls_args << "-t" if args[:t?]
 
```

will result in no error.

Reply:
Perfect answer, thank you for the clear example. I believe the latest commit copies the behavior of method access to hash access, PTAL.

Reply:
I think that's a fair compromise for now. It'd be nice to make the syntax more consistent at some point but that will likely require a more involved refactor.